### ISSUE TITLE
[SOL-38491] Manual ACK / REJECT / REQUEUE not working properly.

### ISSUE THREAD

**Issue by FWinkler79** [SolaceProducts/solace-spring-cloud#14]
I am using the following code to handle manual acknowledgement of events:

```java
@StreamListener(JobTriggerEventConsumerBinding.INPUT)
protected void onJobTriggerEvent(org.springframework.messaging.Message<JobExecutionTriggerEvent> message, 
                                 MessageHeaders headers,
                                 @Validated JobExecutionTriggerEvent event) throws InterruptedException { 
  
  StaticMessageHeaderAccessor.getAcknowledgmentCallback(message).noAutoAck();
  new Thread(new Runnable() {
    public void run() {
      Thread.sleep(5000);
      StaticMessageHeaderAccessor.getAcknowledgmentCallback(message).acknowledge(Status.REQUEUE);
    }
  }).start();
}
```

**Expected Behaviour:** The event should be processed on the newly created thread and after 5 seconds should be re-queued and hence re-delivered.

**Observed Behaviour:** The event is acknowledged without being re-queued. The same is true if `Status.REJECT` is used. For some reason, the event can only be acknowledged.

Looking at `JCSMPAcknowledgementCallback` (which implements the SCSt `AcknowledgementCallback`), one can observe the following code:

```java
JCSMPAcknowledgementCallback(XMLMessage xmlMessage) {
  this.xmlMessage = xmlMessage;
}

@Override
public void acknowledge(Status status) {
  switch (status) {
    case ACCEPT:
      xmlMessage.ackMessage();
      break;
    case REJECT:
      xmlMessage.ackMessage();
      break;
    case REQUEUE:
      xmlMessage.ackMessage();
  }

  acknowledged = true;
}
```
Note, how the message is acknowledged no matter what `case` of the `switch` statement is active.

This defect (if it is one) currently makes it impossible to handle concurrent processing of events on application level, as the acknowledgement / rejection / re-queue would have to be manually handled on each separate processing thread. This is currently not possible.

**Comment by Nephery** [comment]
Switching this to enhancement label since the binder wasn't designed with client-acks in mind.

Also, note that to implement this feature, we would also require for a new consumer config property to be created which will allow for the binder to skip auto-acking messages. But I do have some concerns about how implementing this would bypass some of the error-handling protections that both the binder and SCSt framework provides (namely retries and error-channel redirects). Though this might be fine if we properly document it.

**Comment by FWinkler79** [comment]
Hi @Nephery,

> [...] the binder wasn't designed with client-acks in mind.

If that is the case, then that means that it not able to support [Polled Consumers](https://cloud.spring.io/spring-cloud-static/spring-cloud-stream/current/reference/html/spring-cloud-stream.html#spring-cloud-streams-overview-using-polled-consumers) which seems to be a standard feature of the Spring Cloud Stream API.

In there (and that not only applies to Polled Consumers) it says:

> Normally, the poll() method acknowledges the message when the MessageHandler exits. If the method exits abnormally, the message is rejected (not re-queued), but see Handling Errors. You can override that behavior by taking responsibility for the acknowledgment, as shown in the following example...

The idea here is that unless you enable manual acknowledgement in code the SCSt stack will handle it for you. In that case you can cause rejection / re-queue by throwing exceptions from your handler code. No exception (or exception types that are ignored) in the event handler code implies evth. went fine and leads to auto-acknowledgement.

However, manually handling this is part of the SCSt API, too (as described in the linked documentation above). You enable it in code using `StaticMessageHeaderAccessor.getAcknowledgmentCallback(m).noAutoAck();` - so I do not see why you would need an extra consumer config parameter. It's basically a manual override in code.

**Comment by Mrc0113** [comment]
Hi @FWinkler79 & @Nephery, 

I agree with @FWinkler79 on this piece...manual override in code should be good 👍 :
> However, manually handling this is part of the SCSt API, too (as described in the linked documentation above). You enable it in code using StaticMessageHeaderAccessor.getAcknowledgmentCallback(m).noAutoAck(); - so I do not see why you would need an extra consumer config parameter. It's basically a manual override in code.

As for pollable consumers I agree the binder's support of _Polled Consumers_ will be better once the manual ack is available but I believe right now it _should just work_ with auto acknowledgements upon a successful exit of the MessageHandler which a lot of simpler microservices will likely use . I haven't had a chance to test this to be sure though. It hasn't been a common use case for users of our binder so far. 
> If that is the case, then that means that it not able to support Polled Consumers which seems to be a standard feature of the Spring Cloud Stream API.

@FWinkler79 - just a heads up that I'm hoping to figure out some priorities and rough time frames for this + other binder enhancements soon and will share once I have rough estimates for what's next + time frames.

**Comment by Nephery** [comment]
> As for pollable consumers I agree the binder's support of Polled Consumers will be better once the manual ack is available but I believe right now it should just work with auto acknowledgements upon a successful exit of the MessageHandler which a lot of simpler microservices will likely use . I haven't had a chance to test this to be sure though. It hasn't been a common use case for users of our binder so far.

To add to this, we do [provide the framework with the same error message handler that we provide regular consumers](https://github.com/SolaceProducts/solace-spring-cloud/blob/69586d719162fafd151c67450bce871e904a0c2d/solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder/src/main/java/com/solace/spring/cloud/stream/binder/SolaceMessageChannelBinder.java#L133). So it _should_ react in a similar way as regular consumers in regard to handling exceptions. [We have a test for this](https://github.com/SolaceProducts/solace-spring-cloud/blob/69586d719162fafd151c67450bce871e904a0c2d/solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder/src/test/java/com/solace/spring/cloud/stream/binder/SolaceBinderBasicTest.java#L193).

As for what we're going to do for the actual implementation, my initial thoughts are that properly implementing `REQUEUE` _should_ be straight-forward enough (basically the fix for #13 ), but when we kick out the details for this, we will need to discuss how to deal with `REJECT` status. We will also need to figure out how the acknowledgement callback will play together with our existing error-handling infrastructure as well SCSt's (specifically [the bit how the framework forwards errors to the error channel](https://github.com/SolaceProducts/solace-spring-cloud/blob/master/solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/inbound/JCSMPInboundChannelAdapter.java#L192)). The latter of which, we likely could draw inspiration from the Kafka and Rabbit binders.

**Comment by liuz7** [comment]
when will the manual acknowledgment be ready? Our business logic need consumer to send the manual acknowledgment to broker to re-queue the message if any business failure.

**Comment by GreenRover** [comment]
I see there more than one problem at this point.

See:
com.solace.spring.cloud.stream.binder.util.SolaceErrorMessageHandler#handleMessage
```java
		....
		if (consumerProperties.getExtension().isAutoBindDmq()) {
			republishToDMQ(rawMessage);
		} else if (consumerProperties.getExtension().isRequeueRejected()) {
			requeue(rawMessage);
		} else {
			logger.info(String.format("Raw %s %s: Will be rejected",
					XMLMessage.class.getSimpleName(), rawMessage.getMessageId()));
		}

		AckUtils.autoNack(acknowledgmentCallback); --> this will be execute the next code block
```

com.solace.spring.cloud.stream.binder.util.JCSMPAcknowledgementCallbackFactory.JCSMPAcknowledgementCallback#acknowledge
```java
		public void acknowledge(Status status) {
			switch (status) {
				case ACCEPT:
					xmlMessage.ackMessage();
					break;
				case REJECT:
					xmlMessage.ackMessage();
					break;
				case REQUEUE:
					xmlMessage.ackMessage();
			}

			acknowledged = true;
		}
```

### Cases for messages that are not processable

**Case 1:**   if isAutoBindDmq == true
Messages will be republished into a new queue. Works as designed.
But specify a global DMQ would be nice. If there is a process consuming the messages form dmq to mail those is very complicated with many dmq.

**Case 2:**   if isRequeueRejected() == true
Messages will be republished into the same queue where it was read from. 
The messages will be in the queue for ever. Because the solace broker DMQ retry mechanism will never come to fruition. Because those are always new messages.
Here we normally should restart the `FlowReceiver` to NACK the message.
**Problem**: Is there really a use case, to act like this?

**Case 3:**  none of both options
A log message occurs and the message will be acknowledged. The message is lost.

##########

Conclusion:
The solace DMQ mechanism was undermined.

**Comment by GreenRover** [comment]
@FWinkler79
Using threads in message consumer is generally a bad idea. Because this will destroy the spring retry template.
For the same reason as above, Flux should be used with care.  
Better use: "concurrency" to have scale out in app.

```
spring:
  cloud:
    function:
      definition: receiveTemperatureSensorSimple
    stream:
      bindings:
        receiveTemperatureSensorSimple-in-0:
          destination: sensor/temperature
          group: sensor.temperature
          contentType: "application/json"
          binder: main_session
          consumer:
            defaultRetryable: true # Default: true; Whether exceptions thrown by the listener that are not listed in the retryableExceptions are retryable.
            retryableExceptions:
              javax.validation.ConstraintViolationException: false #Not necessary because defaultRetryable: false
              java.lang.IllegalArgumentException: false
            maxAttempts: 10
            backOffInitialInterval: 100
            backOffMaxInterval: 3_000 # reached at attempt 6, intervals: 0,1s, 0,2s, 0,4s, 0,8s, 1,6s, 3k, 3k, 3k, 3s, 3s == 18,1s
            backOffMultiplier: 2.0
            concurrency: 4 # will create 4 threads, each having a own flow but 1 shared connection
```

**Comment by Nephery** [comment]
@GreenRover All that logic is currently being refactored as part of the work for manual acknowledgments (this issue): https://github.com/SolaceDev/solace-spring-cloud/tree/client-ack . Most of the issues you've listed have already been addressed in that branch.

> Case 1: if isAutoBindDmq == true
Messages will be republished into a new queue. Works as designed.
But specify a global DMQ would be nice. If there is a process consuming the messages form dmq to mail those is very complicated with many dmq.

I recommend opening a new issue for a global DMQ. Also note that this concept is soon to be renamed to "error queue" to avoid the naming conflict with the Solace broker's DMQ.

> Case 2: if isRequeueRejected() == true
Messages will be republished into the same queue where it was read from.
The messages will be in the queue for ever. Because the solace broker DMQ retry mechanism will never come to fruition. Because those are always new messages.
Here we normally should restart the FlowReceiver to NACK the message.
Problem: Is there really a use case, to act like this?

This has already been addressed in the `client-ack` branch. It will properly rebind the flow instead of republishing the message on the queue. Refer to the updated [JCSMPAcknowledgementCallbackFactory](https://github.com/SolaceDev/solace-spring-cloud/blob/client-ack/solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/util/JCSMPAcknowledgementCallbackFactory.java) and the [SolaceErrorMessageHandler](https://github.com/SolaceDev/solace-spring-cloud/blob/client-ack/solace-spring-cloud-stream-binder/solace-spring-cloud-stream-binder-core/src/main/java/com/solace/spring/cloud/stream/binder/util/SolaceErrorMessageHandler.java).

> Case 3: none of both options
A log message occurs and the message will be acknowledged. The message is lost.

With the exception of consumers using anonymous groups, this case was removed in the `client-ack` branch. Requeuing will be the new default error handling strategy.

> Question:
Why was the earlier exiting NACK feature is not longer supported by the solace broker?

I imagine you're referring to [XMLMessage.rejectMessage()](https://docs.solace.com/API-Developer-Online-Ref-Documentation/java/com/solacesystems/jcsmp/XMLMessage.html#rejectMessage()) function. In which case, I don't know what that earlier existing NACK feature did, but our protocol never supported the concept of anything other than deleting messages. That function was deprecated in 4.0, which means since July 2007 (over 10 years ago). Nowadays that's just ancient leftover code.

**Comment by FWinkler79** [comment]
> @FWinkler79
> Using threads in message consumer is generally a bad idea. Because this will destroy the spring retry template.
> For the same reason as above, Flux should be used with care.

Thanks for the remark @GreenRover! 
I surely did not want to imply that using Threads is a good idea here. This was just some quickly assembled code to show the root of the problem.
Ideally, I would not want to start any concurrency in the handler itself, since that already is running concurrently. Most often you probably would not only undermine the retry template by doing so, but also the entire reason for queuing your "packages of work" in a broker in the first place. 
So I totally agree with you.

**Comment by Nephery** [comment]
Closed with #75
